### PR TITLE
xdbg

### PR DESCRIPTION
**Pull request by insipx**
Bulk Identity, Group, and Message generation

```
Debug & Generate data on the XMTP Network

Usage: xdbg [OPTIONS] [COMMAND]

Commands:
  generate  Generate Groups/Messages/Users
  inspect   Inspect Payloads
  query     Query for Information about a Group or Message or User
  info
  export    Export information to JSON
  help      Print this message or the help of the given subcommand(s)

Options:
      --json               Output libxmtp logs into file with a structured, ndJSON format
      --logfmt             Output libxmtp into file with logfmt format
      --human              Output libxmtp logs to file in a human-readable format
  -s, --show-fields        Show key-value fields. Default on for JSON & logfmt and off for human
  -v, --verbose...         Increase logging verbosity
  -q, --quiet...           Decrease logging verbosity
  -b, --backend <BACKEND>  [default: local] [possible values: dev, production, local]
  -u, --url <URL>          URL Pointing to a backend. Conflicts with `backend`
      --clear              Clear ALL local app data & state kept by xdbg Runs at the end of execution, so operations will still be carried out
  -h, --help               Print help
```

# How to run
- install rust
- clone the repository `git clone --single-branch --branch insipx/xmtp-debug git@github.com:xmtp/libxmtp.git`
    - if cloned checkout `insipx/xmtp-debug`
- cd to `xmtp-debug`
- run `cargo xdbg --help` to access the above help message

### Generate 1000 identities to a local node
```bash
cargo xdbg generate --entity identity --amount 1000
```

### Save libxmtp logs from generation event to a file
```bash
cargo xdbg --human generate --entity identity --amount 1000
```

### Generate Groups using randomized identities as the owner
```bash
cargo xdbg generate --entity group --amount 100
```

### Generate Groups using randomized identities as the owner, with randomized identities as members (currently buggy)
```bash
cargo xdbg generate --entity group --amount 100 --invite 10
```

### Generate Identities and then immediately delete all saved app data
```bash
cargo xdbg --clear generate --entity group --amount 100 --invite 10
```

### Delete all saved app data
```bash
cargo xdbg --clear
```

### Show information (# of identities/groups/messages/size of sqlite databases)
```bash
cargo xdbg info
```

LIST:
- [x] Identity Generation
- [x] Group Generation
    - works but has issues with inviting other random members, hitting a PRAGMA salt error and address not found errors, which points to some issues with registration in the CLI
- [x] message generation
    - not yet started
- network inspection
    - not started, but the CLI example app has introspection currently

### CODE REVIEW

**Comment by mkysel** [comment]
will this work with V4/decentralized node out of the box? Or should I add some task to extend this once its in?

**Comment by insipx** [comment]
I'll probably just copy over the Box<dyn Client> stuff from the CLI example app and then it should work out the box

File: xmtp_mls/src/storage/encrypted_store/sqlcipher_connection.rs [hunk 1]
```diff
@@ -228,14 +228,15 @@ impl super::native::ValidatedConnection for EncryptedConnection {
...
+/*
+        if tracing::enabled!(tracing::Level::DEBUG) {
             conn.batch_execute("PRAGMA cipher_log = stderr; PRAGMA cipher_log_level = INFO;")
                 .ok();
         } else {
+*/
             conn.batch_execute("PRAGMA cipher_log = stderr; PRAGMA cipher_log_level = WARN;")
                 .ok();
-        }
+        // }
```
Comment (richardhuaaa) [inline xmtp_mls/src/storage/encrypted_store/sqlcipher_connection.rs:239]: Are these changes deliberate?

**Review by richardhuaaa — APPROVED** [review]
This looks like it'll be super useful.

Any reason the crate `xmtp-debug` is dash-cased rather than snake cased?

File: xmtp_mls/src/storage/encrypted_store/sqlcipher_connection.rs (hunk 1 above)
Comment (insipx) [reply xmtp_mls/src/storage/encrypted_store/sqlcipher_connection.rs:239]: Yeah the SQLCipher logs are a little annoying when generating lots of groups, it just spits out that the logs were enabled so I commented it out temporarily
